### ISSUE TITLE
Wrong error message when providing a database option on 3.5

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-spark-connector#199]
See example.  This works without the "database" option.  With, it fails with the incorrect error message.

Very important: this is a 3.5 cluster I'm attaching to.  So the use of the database option is *wrong*.   We should either ignore the option (if that's possible) or fail with a better message than this.

<img width="1058" alt="Screen Shot 2020-09-28 at 8 12 11 AM" src="https://user-images.githubusercontent.com/5471710/94430717-5b975800-0162-11eb-900d-14d5e6531c55.png">

**Comment by moxious** [comment]
This may be fixed by surfacing driver / session errors.

This query can't succeed because it doesn't make sense (3.5 isn't multidatabase)

Check to see if session throws an error (I bet it does) -- if this error is properly surfaced to the user, then you're done.

**Comment by utnaf** [comment]
I think #183 will fix this as well. 

But when we add the option consistency check this should be solved anyways.

**Comment by moxious** [comment]
We think this is fixed due to previous PR but I need to verify to close.

**Comment by moxious** [comment]
@utnaf this is not fixed by the recent PR merge.   Here's another replication case, now with a different error message.

The endpoint is 3.5.11 causal cluster with 3 core & 1 read-replica.  I can create a replication environment if you need.

```
spark.read.format("org.neo4j.spark.DataSource").option("url", "neo4j://REMOVED:7687")
    .option("authentication.type", "basic")
    .option("authentication.basic.username", "neo4j")
    .option("authentication.basic.password", "(REMOVED)")
    .option("database", "foo")
    .option("query", "MATCH (h:Household) RETURN id(h)").load().show()
```

This code provides a correct result *without* the database option, but with the database option, provides this error:

Result:

```
java.lang.IllegalArgumentException: Please provide a valid READ query
  at org.neo4j.spark.util.ValidationUtil$.isTrue(ValidationUtil.scala:22)
  at org.neo4j.spark.util.Validations$$anonfun$4.apply(Validations.scala:113)
  at org.neo4j.spark.util.Validations$$anonfun$4.apply(Validations.scala:93)
  at org.neo4j.spark.reader.Neo4jDataSourceReader$$anonfun$1.apply(Neo4jDataSourceReader.scala:22)
  at org.neo4j.spark.reader.Neo4jDataSourceReader$$anonfun$1.apply(Neo4jDataSourceReader.scala:22)
  at org.neo4j.spark.Neo4jOptions.validate(Neo4jOptions.scala:161)
  at org.neo4j.spark.reader.Neo4jDataSourceReader.<init>(Neo4jDataSourceReader.scala:22)
  at org.neo4j.spark.DataSource.createReader(DataSource.scala:17)
  at org.neo4j.spark.DataSource.createReader(DataSource.scala:13)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$SourceHelpers.createReader(DataSourceV2Relation.scala:155)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:172)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:204)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:167)
  ... 45 elided
```

**Comment by moxious** [comment]
@utnaf replicate with this branch: https://github.com/utnaf/spark-connector-notebooks/tree/neo4j-3.5 .  I won't PR a 3.5 version of the notebooks, but if you start this remember to remove your neo4j 4 data files before launching

Replication does not require a cluster

**Comment by utnaf** [comment]
Investigating on this I found out that the correct exception is thrown, but it's then caught  and logged. 
This because the first point of the code where we do a connection is the `isValidQuery` method, where we just need to know if the query is runnable or not, the code looks like this:

```
  def isValidQuery(query: String, expectedQueryTypes: org.neo4j.driver.summary.QueryType*): Boolean = try {
    val queryType = session.run(s"EXPLAIN $query").consume().queryType()
    expectedQueryTypes.size == 0 || expectedQueryTypes.contains(queryType)
  } catch {
    case e: Throwable => {
      log.error("Query not compiled because of the following exception:", e)
      false
    }
  }
```

This was the exception is "hidden" in the logs:
```
ERROR [2020-10-19 12:31:33,414] ({FIFOScheduler-interpreter_1552617247-Worker-1} SchemaService.scala[isValidQuery]:407) â Query not compiled because of the following exception:
org.neo4j.driver.exceptions.ClientException: Database name parameter for selecting database is not supported in Bolt Protocol Version 3.0. Database name: 'db1'
```

Possible solutions:
- doing a connection check _before_ everything else
- modify this method to rethrow connection exceptions

**Comment by conker84** [comment]
I vote for `doing a connection check before everything else` so we can also address #170

**Comment by utnaf** [comment]
> I vote for `doing a connection check before everything else` so we can also address #170

This would actually work also for `labels` and `relationship` modes, unlike the second option.

**Comment by moxious** [comment]
I vote the same as @conker84 kill two birds with one stone

**Comment by conker84** [comment]
closed via 60651a3
